### PR TITLE
Add save to case saves to form summary

### PR DESCRIPTION
**Pull request by proteusvacuum**
This adds an extra "node" for each save to case question which then shows which case properties are being updated by that particular question. 

Unfortunately Vellum only sends back the case updates in "bulk" under this root node, otherwise I would have wanted to add this information under each sub node. 

It also changes the icon for these questions to be a diskette, like we have elsewhere. 

screenshot from 2019-02-25 17-22-36

The second commit does a bit of refactoring.

### CODE REVIEW

File: corehq/apps/app_manager/util.py [hunk 1]
```diff
@@ -507,84 +508,102 @@ def get_sort_and_sort_only_columns(detail, sort_elements):
...
+        self.app = app
+        self.include_shadow_forms = include_shadow_forms
+
+        self.errors = []
+
+        self._seen_save_to_case = defaultdict(list)
+        self._case_meta = self.app.get_case_metadata()
+
+    def get_app_data(self):
+        return [self._compile_module(module) for module in self.app.get_modules()], self.errors
```
Comment (esoergel) [inline corehq/apps/app_manager/util.py]: One nitpick with this is `compile_x`, `get_x_data`, and `x.to_json` are all used in this class for essentially the same thing - would be nice to align that terminology.

File: corehq/apps/app_manager/util.py [hunk 2]
```diff
@@ -507,84 +508,102 @@ def get_sort_and_sort_only_columns(detail, sort_elements):
...
+
+    def _compile_module(self, module):
+        return {
             'id': module.unique_id,
             'name': module.name,
             'short_comment': module.short_comment,
             'module_type': module.module_type,
             'is_surveys': module.is_surveys,
             'module_filter': module.module_filter,
+            'forms': [self._compile_form(form) for form in self._get_pertinent_forms(module)],
```
Comment (esoergel) [inline corehq/apps/app_manager/util.py]: I especially like this, where you split up getting the forms and interpreting them

File: corehq/apps/app_manager/util.py [hunk 3]
```diff
@@ -507,84 +508,102 @@ def get_sort_and_sort_only_columns(detail, sort_elements):
...
+        response['save_properties'] = self._case_meta.get_save_properties(form_unique_id, question_path)
+        self._seen_save_to_case[form_unique_id].append(question_path)
+        return response
 
-        module_meta['forms'] = forms
-        modules.append(module_meta)
-    return modules, errors
+
+def get_form_data(domain, app, include_shadow_forms=True):
+    return FormDataGenerator(domain, app, include_shadow_forms).get_app_data()
```
Comment (esoergel) [inline corehq/apps/app_manager/util.py]: This pattern (a class that only exists for its one public method) is a little ugly, so I like that you've kept that wrapped with a simple function and constrained knowledge of the class to this file.

**Review by esoergel — COMMENTED** [review]
(Commenting solely on https://github.com/dimagi/commcare-hq/pull/23409/commits/780c8274840187e4790a5f5d24c8d3742055b598)
This change is a strong positive in my book - it breaks down a very large function into comprehensible components that do about what you'd expect.  I especially like the drilldown style organization.  It would be nice to do this without mutating class attributes during the iteration, though I'm not sure any alternative would end up cleaner.

Probably my biggest issue with these very large functions (the old code) is that you often don't know at a glance what the dependencies for each block of code is or what they might be mutating.  This does a good job of compartmentalizing those bits of scope.

File: corehq/apps/app_manager/util.py [hunk 4]
```diff
@@ -507,84 +508,102 @@ def get_sort_and_sort_only_columns(detail, sort_elements):
...
+            'short_comment': form.short_comment,
+            'action_type': form.get_action_type(),
+            'form_filter': form.form_filter,
+        }
+        try:
+            form_meta['questions'] = [
+                compiled_question
+                for question in form.get_questions(self.app.langs, include_triggers=True,
+                                                   include_groups=True, include_translations=True)
+                for compiled_question in self._compile_form_questions(form.unique_id, question)
```
Comment (millerdev) [inline corehq/apps/app_manager/util.py]: `self._compile_form_questions()` always yields a single item. What do you think of changing it to `return` rather than `yield`, and rephrase this as
```py
form_meta['questions'] = [
    self._compile_form_question(form.unique_id, question)
    for question in form.get_questions(
        self.app.langs, include_triggers=True,
        include_groups=True, include_translations=True)
]
```
?

File: corehq/apps/app_manager/util.py [hunk 5]
```diff
@@ -507,84 +508,102 @@ def get_sort_and_sort_only_columns(detail, sort_elements):
...
-                }
-                form_meta['module'] = copy(module_meta)
-                errors.append(form_meta)
-            else:
-                forms.append(form_meta)
+            self.errors.append(form_meta)
+        return form_meta
+
+    def _compile_form_questions(self, form_unique_id, question):
+        from corehq.apps.reports.formdetails.readable import FormQuestionResponse
```
Comment (millerdev) [inline corehq/apps/app_manager/util.py]: nit: is there a reason this can't be moved to the top of the file? (same in `_get_save_to_case_root_node`)

**Review by millerdev — COMMENTED** [review]
Nice refactor.

File: corehq/apps/app_manager/util.py (hunk 4 above)
Comment (proteusvacuum) [reply corehq/apps/app_manager/util.py]: Thanks for this. It actually is supposed to yield 2 items if the question requires a new save to case block to be added. I've updated that in a later commit.

File: corehq/apps/app_manager/util.py (hunk 5 above)
Comment (proteusvacuum) [reply corehq/apps/app_manager/util.py]: Good point. This was causing an import loop, which suggests to me that this was in the wrong place. I've moved this whole class alongside the other code that does the same things here: https://github.com/dimagi/commcare-hq/pull/23409/commits/e55e7832defb68b81e08160b3a7bb94c336fc524

File: corehq/apps/app_manager/util.py [hunk 6]
```diff
@@ -507,84 +508,102 @@ def get_sort_and_sort_only_columns(detail, sort_elements):
...
+        question_path = question['value'].split('/case/')[0]
+        if is_save_to_case and question_path not in self._seen_save_to_case[form_unique_id]:
+            yield self._get_save_to_case_root_node(form_unique_id, question)
+        else:
+            response = FormQuestionResponse(question).to_json()
+            response['load_properties'] = self._case_meta.get_load_properties(form_unique_id, question['value'])
+            response['save_properties'] = self._case_meta.get_save_properties(form_unique_id, question['value'])
+            if is_save_to_case:
+                response['type'] = 'SaveToCase'
+            yield response
```
[the remark this answers is not in the document]
Comment (proteusvacuum) [reply corehq/apps/app_manager/util.py]: good point. I've extracted a few more methods from this to make it a bit clearer.

**Review by esoergel — APPROVED** [review]
lookin good

**Review by millerdev — APPROVED** [review]
